### PR TITLE
refactor: account for built-in oracle to check on prices during a collect

### PR DESCRIPTION
**Pull request by kupermind**
- Account for built-in oracle to check on prices during a collect

### CODE REVIEW

File: contracts/MemeFactory.sol [hunk 1]
```diff
@@ -201,11 +202,72 @@ abstract contract MemeFactory {
...
+        uint256 twapPrice = _getTwapFromOracle(factory, token0, token1, 1800);
+        uint256 instantPrice = (uint256(sqrtPriceX96) * uint256(sqrtPriceX96)) / (1 << 192);
+
+        uint256 deviation = (instantPrice > twapPrice) ?
+            ((instantPrice - twapPrice) * 1e18) / twapPrice :
+            ((twapPrice - instantPrice) * 1e18) / twapPrice;
+
+        // TODO How to calculate it?
+        uint256 maxAllowedDeviation = 10;
+        require(deviation <= maxAllowedDeviation, "Price deviation too high");
```
Comment (77ph) [inline contracts/MemeFactory.sol:261]: maxAllowedDeviation * 10^16? if maxAllowedDeviation in %
or 
maxAllowedDeviation  in basispoint like 10000 == 100%

File: contracts/MemeFactory.sol [hunk 2]
```diff
@@ -201,11 +202,72 @@ abstract contract MemeFactory {
...
+        // 30-minute TWAP
+        uint256 twapPrice = _getTwapFromOracle(factory, token0, token1, 1800);
+        uint256 instantPrice = (uint256(sqrtPriceX96) * uint256(sqrtPriceX96)) / (1 << 192);
+
+        uint256 deviation = (instantPrice > twapPrice) ?
+            ((instantPrice - twapPrice) * 1e18) / twapPrice :
+            ((twapPrice - instantPrice) * 1e18) / twapPrice;
+
+        // TODO How to calculate it?
+        uint256 maxAllowedDeviation = 10;
```
Comment (DavidMinarsch) [inline contracts/MemeFactory.sol:260]: Let's make it a constructor argument. I suggest it should be smaller than 1%. All we are doing is that there wasn't much volatility in the last 30 min. We are not actually causing any slippage ourselves.

File: contracts/MemeFactory.sol [hunk 3]
```diff
@@ -201,11 +202,72 @@ abstract contract MemeFactory {
...
+        // Verify pool reserves before proceeding
+        address pool = IUniswapV3(factory).getPool(token0, token1, FEE_TIER);
+        require(pool != address(0), "Pool does not exist");
+
+        // Get current pool reserves
+        (uint160 sqrtPriceX96, , , , , , ) = IUniswapV3(pool).slot0();
+
+        // Check TWAP or historical data
+        // 30-minute TWAP
+        uint256 twapPrice = _getTwapFromOracle(factory, token0, token1, 1800);
```
Comment (DavidMinarsch) [inline contracts/MemeFactory.sol:252]: just pass in the pool to avoid double retrival

File: contracts/MemeFactory.sol [hunk 4]
```diff
@@ -201,11 +202,72 @@ abstract contract MemeFactory {
...
+        uint256 maxAllowedDeviation = 10;
+        require(deviation <= maxAllowedDeviation, "Price deviation too high");
+    }
+
     /// @dev Collects fees from LP position, burns the meme token part and schedules for ascendance the native token part.
     /// @param memeToken Meme token address.
     /// @param positionId LP position ID.
     /// @param isNativeFirst Order of a native token in the pool.
     function _collectFees(address memeToken, uint256 positionId, bool isNativeFirst) internal {
+        _checkPoolPrices(memeToken, isNativeFirst);
```
Comment (77ph) [inline contracts/MemeFactory.sol:269]: check do nothing - just return unused bool

Comment (77ph) [inline contracts/MemeFactory.sol:269]: loss latest require(), sorry. all good

File: contracts/MemeFactory.sol (hunk 2 above)
Comment (kupermind) [reply contracts/MemeFactory.sol]: We'll leave as constant, no difference. The difficulty with lots of construct args is to manage them.

**Pull request merged**